### DISCUSSION TITLE
Running with non-global forcing data

### DISCUSSION THREAD

**Discussion by leiapauline** [NCAR/hrldas#186] [Q&A] [answered]
Hello,

I want to run hrldas with my forcing data over 85-165E, -15,35N. I want to reduce the space needed for my input data because our study area is only over a city. What I did was to subset the Zfile and LANDSfile over the same domain as my forcing data. But when I run ./create_forcing.exe, I get an error stating that the z variable can't be read. I checked their structures but I couldn't see any difference. Is there a way for me to make this run possible? I am currently using hrldas on a docker desktop for windows. Any guidance will be much appreciated.

Sincerely,
Leia Pauline

**Comment by cenlinhe** [comment]
what forcing dataset are you using? ERA5-land? GLDAS? or something else?

**Comment by leiapauline** [comment]
I am using ERA5-Land. If it can help, my input files can be found [here](https://drive.google.com/drive/folders/121vepTsVtZqIhcOPu1qNwMyqaj3ZXXoF?usp=sharing).

**Comment by tslin2** [comment, marked as answer]
Hi,

You will have to revise the subroutine code for the regional files, and recompile it (create_forcing.F) due to the default code being read Zfile and LANDSfile as the global domain.

https://github.com/NCAR/hrldas/blob/087f1fd78601b37eaeead93da9f30253441f7cc2/hrldas/HRLDAS_forcing/create_forcing.F#L2090

You should change these two parts at least.
https://github.com/NCAR/hrldas/blob/087f1fd78601b37eaeead93da9f30253441f7cc2/hrldas/HRLDAS_forcing/create_forcing.F#L2100
https://github.com/NCAR/hrldas/blob/087f1fd78601b37eaeead93da9f30253441f7cc2/hrldas/HRLDAS_forcing/create_forcing.F#L2135

**Comment by leiapauline** [comment]
Thanks for the guidance. The recompiled version is running.

**Reply by fdesales** [reply]
Hi there.  
I was wondering if you had success in running HRLDAS using non-global ERA5 forcing data.  I have edited `create_forcing.F` following the instructions above and recompiled the code. However, my forcing data created with the non-global data looks different than the same generated from global data.  Any advice?  Thanks.

**Reply by cenlinhe** [reply]
Hi, have you checked the lat,lon setup in your modified script. The setup of lon of 0-360deg is different from -180 - 180deg. Also, make sure you select the correct number of nlon and nlat. What study domain are you looking at?

**Reply by fdesales** [reply]
Thanks your response.  I did update nlon and nlat according to my data but that did not solve the problem.  `create_forcing.exe` runs fine but when I compared the **forcing** files generated with global and non-global data, I noticed that the latter had the latitude reversed.  

The problem seem to be in the grib1_gribdata (module_grib1.F).  Global grib data are read by the GRIB1_SGUP_BITMAP subroutine, while for non-global files GRIB1_SGUP_NOBITMAP is used.  The latter flips the latitudes of the data.

My conclusion is that global and non-global ERA5-land grib files downloaded from Copernicus CDS have different BITMAPs.   Maybe this is an issue with the new CDS.

**Reply by scappsphd** [reply]
I am also trying to use regional ERA5 data and have followed these instructions:

I have subsetted the ERA5 data and geopotential and lsm invariant files as follows:
  Lat (y) 26 to 53 degrees North
  Lon (x) 234 to 260 degrees East (-126 degrees and -100 degrees)
  "area": [53, -126, 26, -100]

In create_forcing.F, I have updated the following line 2100:
integer, parameter :: NLATS = 271, NLONS = 261

And, line 2136:
call map_set(PROJ_LATLON, datastruct%proj, lat1=53.00, lon1=234.00, &
         latinc=-0.1, loninc=0.1, knowni=1.0, knownj=1.0)

When running create_forcing.exe, I am receiving this error after the first timestep is read in:

 _You have this many undefined T points over valid land points:         65025
 This tends to happen around coastlines or for unresolved islands in your forcing data
 ===== Consider increasing expand_loop in namelist =====
 However, this may also occur if your forcing source does not cover your domain.
Warning: ieee_inexact is signaling
FORTRAN STOP_

This is due to undefined T over land.  I have checked the data and the input grib data looks ok.  Any suggestions?

**Reply by cenlinhe** [reply]
This is due to the spatial interpolation process in create_forcing that does not resolve some coastal pixels or small island pixels. please increase the expand_loops value in your namelist to 20 or 30 to see if it works.

**Reply by cenlinhe** [reply]
also make sure your forcing data region is larger than your target study domain.

**Reply by scappsphd** [reply]
Thank you for your quick response.  

Yes, my forcing data extent is larger than the extent of my geo_em.d04.nc domain.
And, I tried changing the expand_loops from 10 to 30 and I still get the same error message.

One thing I noticed is that the longitude array for the geopotential file is ordered differently than the forcing files:

The forcing file:
longitude = -126, -125.9, -125.8, -125.7, -125.6, -125.5, -125.4, -125.3, 
    -125.2, -125.1, -125, -124.9, -124.8, -124.7, -124.6, -124.5, -124.4, 
    -124.3, -124.2, -124.1, -124, -123.9, -123.8, -123.7, -123.6, -123.5, 
    -123.4, -123.3, -123.2, -123.1, -123, -122.9, -122.8, -122.7, -122.6,

And, the invariant geopotential file:
longitude = 234, 234.1, 234.2, 234.3, 234.4, 234.5, 234.6, 234.7, 234.8, 
    234.9, 235, 235.1, 235.2, 235.3, 235.4, 235.5, 235.6, 235.7, 235.8, 
    235.9, 236, 236.1, 236.2, 236.3, 236.4, 236.5, 236.6, 236.7, 236.8, 
    236.9, 237, 237.1, 237.2, 237.3, 237.4, 237.5, 237.6, 237.7, 237.8, 
    237.9, 238, 238.1, 238.2, 238.3, 238.4, 238.5, 238.6, 238.7, 238.8,

Both seem to start in the NW corner but one uses longitudes from -180 to 180 and the other uses longitudes 0 to 360 E.  

In the create_forcing.F code, I set as follows:
  integer, parameter :: NLATS = 271, NLONS = 261

  call map_set(PROJ_LATLON, datastruct%proj, lat1=53.00, lon1=-126.00, &
         latinc=-0.1, loninc=0.1, knowni=1.0, knownj=1.0)

But, I may have to change the longitude array for the invariant file to match the forcing file?

**Reply by cenlinhe** [reply]
Yes, the longitude needs to be from -180 to 180

**Reply by scappsphd** [reply]
Still trying to resolve this.  I transformed the longitude variables to range from -180 to 180 in the invariant files using these commands:

ncap2 -O -s 'where(longitude > 180) longitude=longitude-360' geo_1279l4_0.1x0.1.grib2_v4_unpack_westernus.nc geo_1279l4_0.1x0.1.grib2_v4_unpack_westernus-correctedlon.nc
ncap2 -O -s 'where(longitude > 180) longitude=longitude-360' lsm_1279l4_0.1x0.1.grb_v4_unpack_westernus.nc lsm_1279l4_0.1x0.1.grb_v4_unpack_westernus-correctedlon.nc

Unfortunately, this did not resolve the issue.

**Reply by tslin2** [reply]
https://github.com/NCAR/hrldas/discussions/186#discussioncomment-12565843

**Comment by cenlinhe** [comment]
@tslin2 @CharlesZheZhang Could you please take a look at this issue when you have time?

**Reply by tslin2** [reply]
Try either one of the following methods:
* change the longitude from 0 to 360 in forcing files

* change geo_em%lon(i,j)+360 to geo_em%lon(i,j) in this line of code and recompile
 https://github.com/NCAR/hrldas/blob/a4f4a2c6d42d7ad3a11103a18836f822af3f6693/hrldas/HRLDAS_forcing/create_forcing.F#L1272

**Reply by scappsphd** [reply]
Thank you.  I changed line 1272 and it appears to have solved the issue.  

Running HRLDAS with regional data is very important since it avoids downloading excess data we don't need and reduces traffic on servers and networks.  Not to mention, it makes HRLDAS run much faster.  

Your quick response is very much appreciated.  

For documentation and future users, I will reply with a summary of my issue and the fix.

**Reply by fdesales** [reply]
I had the same problem earlier this year.  I can't find my original post, but I think there is a minor issue in the ERA-land GRIB files which causes this inverted-longitude problem when using non-global data as HRLDAS forcing.  Having that noted on a formal HRLDAS manual would be great!
